File: proj/CoreLibrary/src/com/jasonzqshen/familyaccounting/core/masterdata/GLAccountMasterDataFactory.java
package com.jasonzqshen.familyaccounting.core.masterdata;

import java.util.ArrayList;

import org.w3c.dom.Element;

import com.jasonzqshen.familyaccounting.core.CoreDriver;
import com.jasonzqshen.familyaccounting.core.exception.IdentityInvalidChar;
import com.jasonzqshen.familyaccounting.core.exception.IdentityNoData;
import com.jasonzqshen.familyaccounting.core.exception.IdentityTooLong;
import com.jasonzqshen.familyaccounting.core.exception.MasterDataIdentityExists;
import com.jasonzqshen.familyaccounting.core.exception.MasterDataIdentityNotDefined;
import com.jasonzqshen.familyaccounting.core.exception.NoGLAccountGroupException;
import com.jasonzqshen.familyaccounting.core.exception.NullValueNotAcceptable;
import com.jasonzqshen.familyaccounting.core.exception.ParametersException;
import com.jasonzqshen.familyaccounting.core.exception.format.MasterDataFileFormatException;
import com.jasonzqshen.familyaccounting.core.exception.runtime.SystemException;
import com.jasonzqshen.familyaccounting.core.utils.CoreMessage;
import com.jasonzqshen.familyaccounting.core.utils.GLAccountGroup;
import com.jasonzqshen.familyaccounting.core.utils.MessageType;
import com.jasonzqshen.familyaccounting.core.utils.StringUtility;

public class GLAccountMasterDataFactory extends MasterDataFactoryBase {
	/**
	 * 
	 * @param parser
	 * @param coreDriver
	 */
	public GLAccountMasterDataFactory(CoreDriver coreDriver, MasterDataManagement management) {
		super(coreDriver, management);
	}

	@Override
	public MasterDataBase createNewMasterDataBase(MasterDataIdentity identity,
			String descp, Object... objects) throws ParametersException,
			MasterDataIdentityNotDefined, MasterDataIdentityExists {
		// check id is G/L identity
		if (!(identity instanceof MasterDataIdentity_GLAccount)) {
			throw new ParametersException(
					"Type of master data identity should be MasterDataIdentity_GLAccount.");
		}
		// check duplicated
		if (_list.containsKey(identity)) {
			throw new MasterDataIdentityExists();
		}

		MasterDataIdentity_GLAccount identity_gl = (MasterDataIdentity_GLAccount) identity;

		MasterDataIdentity bankAccount = null;

		if (objects.length == 0 || objects.length == 1) {
		} else {
			throw new ParametersException(String.format(
					CoreMessage.ERR_PARAMETER_LENGTH, 1, objects.length));
		}

		if (objects.length == 1) {
			// bank account
			if (!(objects[0] instanceof MasterDataIdentity)) {
				throw new ParametersException(String.format(
						CoreMessage.ERR_PARAMETER_TYPE,
						MasterDataIdentity.class.getName()));
			}
			bankAccount = (MasterDataIdentity) objects[0];
		}

		GLAccountMasterData glAccount;
		try {
			glAccount = new GLAccountMasterData(_coreDriver, _management, identity_gl,
					descp, bankAccount);
		} catch (NullValueNotAcceptable e) {
			throw new SystemException(e);
		} catch (NoGLAccountGroupException e) {
			throw new ParametersException(identity_gl.toString());
		}

		this._containDirtyData = true;
		this._list.put(identity_gl, glAccount);

		// raise create master data
		_coreDriver.getListenersManagement().createMasterData(this, glAccount);
		_coreDriver.logDebugInfo(this.getClass(), 84,
				String.format("Create G/L account (%s).", glAccount.toXML()),
				MessageType.INFO);
		return glAccount;
	}

	@Override
	public MasterDataBase parseMasterData(CoreDriver coreDriver, Element elem)
			throws MasterDataFileFormatException {
		String id = elem.getAttribute(MasterDataUtils.XML_ID);
		String descp = elem.getAttribute(MasterDataUtils.XML_DESCP);

		String bankAccStr = elem.getAttribute(MasterDataUtils.XML_BANK_ACCOUNT);

		try {
			MasterDataIdentity_GLAccount identity = new MasterDataIdentity_GLAccount(
					id.toCharArray());

			GLAccountMasterData glAccount = null;

			// bank account
			if (StringUtility.isNullOrEmpty(bankAccStr)) {
				glAccount = (GLAccountMasterData) this.createNewMasterDataBase(
						identity, descp);
			} else {
				MasterDataIdentity bankAccId = new MasterDataIdentity(
						bankAccStr.toCharArray());
				glAccount = (GLAccountMasterData) this.createNewMasterDataBase(
						identity, descp, bankAccId);

			}

			_coreDriver
					.logDebugInfo(
							this.getClass(),
							167,
							String.format("Parse G/L account (%s).",
									glAccount.toXML()), MessageType.INFO);
			return glAccount;
		} catch (IdentityTooLong e) {
			_coreDriver.logDebugInfo(this.getClass(), 150,
					"Master data identity is too long.", MessageType.ERROR);
			throw new MasterDataFileFormatException(MasterDataType.GL_ACCOUNT);
		} catch (IdentityNoData e) {
			_coreDriver
					.logDebugInfo(this.getClass(), 154,
							"Master data identity is with no value.",
							MessageType.ERROR);
			throw new MasterDataFileFormatException(MasterDataType.GL_ACCOUNT);
		} catch (IdentityInvalidChar e) {
			_coreDriver.logDebugInfo(this.getClass(), 160,
					"Invalid character in identity.", MessageType.ERROR);
			throw new MasterDataFileFormatException(MasterDataType.GL_ACCOUNT);
		} catch (ParametersException e) {
			_coreDriver.logDebugInfo(this.getClass(), 164,
					"Function parameter set error: " + e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		} catch (MasterDataIdentityExists e) {
			_coreDriver.logDebugInfo(this.getClass(), 168,
					"Master data identity duplicated.", MessageType.ERROR);
			throw new MasterDataFileFormatException(MasterDataType.GL_ACCOUNT);
		} catch (MasterDataIdentityNotDefined e) {
			_coreDriver.logDebugInfo(this.getClass(), 173,
					"Identity has not been defined.", MessageType.ERROR);
			throw new MasterDataFileFormatException(MasterDataType.GL_ACCOUNT);
		}
	}

	/**
	 * get balance accounts
	 * 
	 * @return
	 */
	public GLAccountMasterData[] getBalanceAccounts() {
		return getAccounts(GLAccountGroup.BALANCE_GROUP);
	}

	/**
	 * get liquidity accounts
	 * 
	 * @return
	 */
	public GLAccountMasterData[] getLiquidityAccounts() {
		return getAccounts(GLAccountGroup.Liquidity_GROUP);
	}

	/**
	 * get liability account
	 * 
	 * @return
	 */
	public GLAccountMasterData[] getLiabilityAccounts() {
		return getAccounts(GLAccountGroup.LIABILITIES_GROUP);
	}

	/**
	 * get revenue account
	 * 
	 * @return
	 */
	public GLAccountMasterData[] getRevenueAccounts() {
		return getAccounts(GLAccountGroup.REVENUE_GROUP);
	}

	/**
	 * get cost account
	 * 
	 * @return
	 */
	public GLAccountMasterData[] getCostAccounts() {
		return getAccounts(GLAccountGroup.COST_GROUP);
	}

	/**
	 * get G/L accounts base on the group
	 * 
	 * @param groups
	 * @return
	 */
	private GLAccountMasterData[] getAccounts(GLAccountGroup[] groups) {
		ArrayList<GLAccountMasterData> array = new ArrayList<GLAccountMasterData>();
		for (MasterDataBase master : this._list.values()) {
			GLAccountMasterData glAccount = (GLAccountMasterData) master;
			for (int i = 0; i < groups.length; ++i) {
				if (glAccount.getGroup() == groups[i]) {
					array.add(glAccount);
					break;
				}
			}
		}

		GLAccountMasterData[] ret = new GLAccountMasterData[array.size()];
		for (int i = 0; i < array.size(); ++i) {
			ret[i] = array.get(i);
		}
		return ret;
	}
}

File: branches/V1.0/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/MasterData/GLAccountMasterDataFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Linq;
using WinStore_FamilyAccountingCore.Exceptions;
using WinStore_FamilyAccountingCore.Exceptions.FormatExceptions;
using WinStore_FamilyAccountingCore.Utilities;

namespace WinStore_FamilyAccountingCore.MasterData
{
    public class GLAccountMasterDataFactory : MasterDataFactoryBase
    {

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="coreDriver"></param>
        /// <param name="management"></param>
        /// <exception cref="ArgumentNullException">Argument is null</exception>
        public GLAccountMasterDataFactory(CoreDriver coreDriver,
                MasterDataManagement management)
            : base(coreDriver, management)
        {
        }

        /// <summary>
        /// Create new master data 
        /// </summary>
        /// <param name="identity"></param>
        /// <param name="descp"></param>
        /// <param name="?"></param>
        /// <returns></returns>
        /// <exception cref="ParametersException">Parameters Exception</exception>
        /// <exception cref="MasterDataIdentityExists">Duplicated master data identity exists </exception>
        /// <exception cref="MasterDataIdentityNotDefined">Master data identity is not defined</exception>
        /// <exception cref="ArgumentNullException">Argument is null</exception>
        /// <exception cref="SystemException">Bug</exception>
        public override MasterDataBase CreateNewMasterDataBase(MasterDataIdentity identity,
                String descp, params Object[] objects)
        {
            // check id is G/L identity
            MasterDataIdentity_GLAccount identity_gl = identity as MasterDataIdentity_GLAccount;
            if (identity_gl == null)
            {
                throw new ParametersException();
            }
            // check duplicated
            if (_list.ContainsKey(identity))
            {
                throw new MasterDataIdentityExists();
            }

            MasterDataIdentity bankAccount = null;

            if (objects.Length == 0 || objects.Length == 1)
            {
            }
            else
            {
                throw new ParametersException(1, objects.Length);
            }

            if (objects.Length == 1)
            {
                // contain bank account
                bankAccount = objects[0] as MasterDataIdentity;
                if (bankAccount == null)
                {
                    throw new ParametersException();
                }
            }

            GLAccountMasterData glAccount;
            try
            {
                glAccount = new GLAccountMasterData(_coreDriver, _management,
                        identity_gl, descp, bankAccount);
            }
            catch (ArgumentNullException e)
            {
                throw new SystemException(e);
            }
            catch (NoGLAccountGroupException)
            {
                throw new ParametersException();
            }

            this._containDirtyData = true;
            this._list.Add(identity_gl, glAccount);

            // raise create master data
            _coreDriver.ListenerMgmt.CreateMasterData(this, glAccount);
            _coreDriver.logDebugInfo(this.GetType(), 84,
                    String.Format("Create G/L account ({0}).", glAccount.Identity.ToString()),
                    MessageType.INFO);
            return glAccount;
        }

        /// <summary>
        /// Parse master data from XML
        /// </summary>
        /// <param name="coreDriver"></param>
        /// <param name="elem"></param>
        /// <returns></returns>
        /// <exception cref="MasterDataFileFormatException">Master Data file exception</exception>
        /// <exception cref="ArgumentNullException">Argument is null</exception>
        /// <exception cref="SystemException">Bug</exception>
        public override MasterDataBase ParseMasterData(CoreDriver coreDriver, XElement elem)
        {
            XAttribute id = elem.Attribute(MasterDataUtils.XML_ID);
            XAttribute descp = elem.Attribute(MasterDataUtils.XML_DESCP);
            XAttribute bankAccStr = elem.Attribute(MasterDataUtils.XML_BANK_ACCOUNT);
            XAttribute initAmountAttr = elem.Attribute(MasterDataUtils.XML_INIT_AMOUNT);

            MasterDataIdentity_GLAccount identity;
            MasterDataIdentity bankAccId = null;
            try
            {
                identity = new MasterDataIdentity_GLAccount(
                        id.Value);

                // bank account
                if (bankAccStr != null)
                {
                    bankAccId = new MasterDataIdentity(
                            bankAccStr.Value);
                }

            }
            catch (Exception e)
            {
                _coreDriver.logDebugInfo(this.GetType(), 150, e.Message, MessageType.ERRO);
                throw new MasterDataFileFormatException(MasterDataType.GL_ACCOUNT);
            }

            CurrencyAmount initAmount = new CurrencyAmount();
            if (initAmountAttr != null)
            {
                try
                {
                    initAmount = CurrencyAmount.Parse(initAmountAttr.Value);
                }
                catch (CurrencyAmountFormatException e)
                {
                    _coreDriver.logDebugInfo(this.GetType(), 150, e.Message, MessageType.ERRO);
                    throw new MasterDataFileFormatException(MasterDataType.GL_ACCOUNT);
                }
            }
            try
            {
                GLAccountMasterData glAccount;
                if (bankAccId == null)
                {
                    glAccount = (GLAccountMasterData)this.CreateNewMasterDataBase(
                            identity, descp.Value);
                }
                else
                {
                    glAccount = (GLAccountMasterData)this.CreateNewMasterDataBase(
                                identity, descp.Value, bankAccId);
                }
                glAccount.setInitAmountInternal(initAmount);
                _coreDriver
                            .logDebugInfo(
                                    this.GetType(),
                                    167,
                                    String.Format("Parse G/L account ({0}).",
                                            glAccount.Identity.ToString()), MessageType.INFO);
                return glAccount;
            }
            catch (Exception e)
            {
                throw new SystemException(e);
            }

        }

        /// <summary>
        /// balance accounts
        /// </summary>
        /// <returns></returns>
        public List<GLAccountMasterData> BalanceAccounts
        {
            get
            {
                return GetAccounts(GLAccountGroup.BALANCE_GROUP);
            }
        }

        /// <summary>
        /// liquidity accounts
        /// </summary>
        public List<GLAccountMasterData> LiquidityAccounts
        {
            get
            {
                return GetAccounts(GLAccountGroup.Liquidity_GROUP);
            }
        }

        /// <summary>
        /// liability accounts
        /// </summary>
        public List<GLAccountMasterData> LiabilityAccounts
        {
            get
            {
                return GetAccounts(GLAccountGroup.LIABILITIES_GROUP);
            }
        }

        /// <summary>
        /// get revenue accounts
        /// </summary>
        public List<GLAccountMasterData> RevenueAccounts
        {
            get
            {
                return GetAccounts(GLAccountGroup.REVENUE_GROUP);
            }
        }

       /// <summary>
       /// get cost accounts
       /// </summary>
        public List<GLAccountMasterData> CostAccounts{ get { return GetAccounts(GLAccountGroup.COST_GROUP); } }

       /// <summary>
       /// get accounts based on the groups
       /// </summary>
       /// <param name="groups"></param>
       /// <returns></returns>
        public List<GLAccountMasterData> GetAccounts(GLAccountGroupENUM[] groups)
        {
            List<GLAccountMasterData> array = new List<GLAccountMasterData>();
            foreach (var item in this._list)
            {
                GLAccountMasterData glAccount = (GLAccountMasterData)item.Value;
                for (int i = 0; i < groups.Length; ++i)
                {
                    if (glAccount.Group.Identity == groups[i])
                    {
                        array.Add(glAccount);
                        break;
                    }
                }
            }

            array.Sort();

            return array;
        }
    }
}
